src/api/c/sparse.cpp

```diff
@@ -0,0 +1,407 @@
+            default: TYPE_ERROR(1, info.getType());
+        }
+        std::swap(*out, output);
+
+    } CATCHALL;
+
+    return AF_SUCCESS;
+}
+
+template<typename T>
+af_array sparseConvertStorage(const af_array in_, const af_sparse_storage destStorage)
+{
+    const SparseArray<T> in = getSparseArray<T>(in_);
+
+    // Only destStorage == AF_SPARSE_DENSE is supported
+    // All the other calls are for future when conversions are supported in
+    // the backend
+    if(destStorage == AF_SPARSE_DENSE) {
+        // Returns a regular af_array, not sparse
+        switch(in.getStorage()) {
+            case AF_SPARSE_CSR:
+                return getHandle(detail::sparseConvertStorageToDense<T, AF_SPARSE_CSR>(in));
+            case AF_SPARSE_CSC:
+                return getHandle(detail::sparseConvertStorageToDense<T, AF_SPARSE_CSC>(in));
+            case AF_SPARSE_COO:
+                return getHandle(detail::sparseConvertStorageToDense<T, AF_SPARSE_COO>(in));
+            default:
+                AF_ERROR("Invalid storage type of input array", AF_ERR_ARG);
+        }
+    } else if(destStorage == AF_SPARSE_CSR) {
+        // Returns a sparse af_array
+        switch(in.getStorage()) {
+            case AF_SPARSE_CSR:
+                return retainSparseHandle<T>(in_);
+            case AF_SPARSE_CSC:
+                return getHandle(detail::sparseConvertStorageToStorage<T, AF_SPARSE_CSR, AF_SPARSE_CSC>(in));
+            case AF_SPARSE_COO:
+                return getHandle(detail::sparseConvertStorageToStorage<T, AF_SPARSE_CSR, AF_SPARSE_COO>(in));
+            default:
+                AF_ERROR("Invalid storage type of input array", AF_ERR_ARG);
+        }
+    } else if(destStorage == AF_SPARSE_CSC) {
+        // Returns a sparse af_array
+        switch(in.getStorage()) {
+            case AF_SPARSE_CSR:
+                return getHandle(detail::sparseConvertStorageToStorage<T, AF_SPARSE_CSC, AF_SPARSE_CSR>(in));
+            case AF_SPARSE_CSC:
+                return retainSparseHandle<T>(in_);
+            case AF_SPARSE_COO:
+                return getHandle(detail::sparseConvertStorageToStorage<T, AF_SPARSE_CSC, AF_SPARSE_COO>(in));
+            default:
+                AF_ERROR("Invalid storage type of input array", AF_ERR_ARG);
+        }
+    } else if(destStorage == AF_SPARSE_COO) {
+        // Returns a sparse af_array
+        switch(in.getStorage()) {
+            case AF_SPARSE_CSR:
+                return getHandle(detail::sparseConvertStorageToStorage<T, AF_SPARSE_COO, AF_SPARSE_CSR>(in));
+            case AF_SPARSE_CSC:
+                return getHandle(detail::sparseConvertStorageToStorage<T, AF_SPARSE_COO, AF_SPARSE_CSC>(in));
+            case AF_SPARSE_COO:
+                return retainSparseHandle<T>(in_);
+            default:
+                AF_ERROR("Invalid storage type of input array", AF_ERR_ARG);
+        }
+    }
+
+    // Shoud never come here
```

Comment:
I had originally done it as nested functions, but you will need to check for the storage types to be the same in both functions then.